docs/framework/guides/tutorials/data-from-external-source.md

```diff
@@ -0,0 +1,877 @@
+
+```js
+// external-widget/externalwidgetcommand.js
+
+import Command from '@ckeditor/ckeditor5-core/src/command';
+
+// example external data source url
+const RESOURCE_URL = 'https://api2.binance.com/api/v3/ticker/24hr?symbol=BTCUSDT';
+
+class ExternalWidgetCommand extends Command {
+	execute() {
+		const editor = this.editor;
+		const selection = editor.model.document.selection;
+
+		editor.model.change( writer => {
+			// Create an <externalElement> element with the "data-resource-url" attribute
+			// (and all the selection attributes)...
+			const externalWidget = writer.createElement(
+				'externalElement', {
+					...Object.fromEntries( selection.getAttributes() ),
+					'data-resource-url': RESOURCE_URL
+				}
+			);
+
+			// ... insert it into the document and put the selection on the inserted element.
+			editor.model.insertObject( externalWidget, null, null, {
+				setSelection: 'on'
+			} );
+		} );
+	}
+
+	refresh() {
+		const model = this.editor.model;
+		const selection = model.document.selection;
+
+		const isAllowed = model.schema.checkChild( selection.focus.parent, 'externalElement' );
+
+		this.isEnabled = isAllowed;
+	}
+}
+```
+
+Import the created command and add it to the editor commands:
+
+```js
+// external-widget/externalwidgetediting.js
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+
+import { toWidget } from '@ckeditor/ckeditor5-widget/src/utils';
+import Widget from '@ckeditor/ckeditor5-widget/src/widget';
+
+import ExternalWidgetCommand from './externalwidgetcommand';                   // ADDED
+
+import './theme/externalwidget.css';
+
+export default class ExternalWidgetEditing extends Plugin {
+	static get requires() {
+		return [ Widget ];
+	}
+
+	init() {
+		console.log( 'ExternalWidgetEditing#init() got called' );
+
+		this._defineSchema();
+		this._defineConverters();
+
+		// ADDED
+		this.editor.commands.add( 'external', new ExternalWidgetCommand( this.editor ) );
+	}
+
+	_defineSchema() {
+		// ...
+	}
+
+	_defineConverters() {
+		// ...
+	}
+}
+```
+
+## Creating the UI
+
+The UI part will provide a button that user can click to insert external widget into the editor.
+
+```js
+// external-widget/externalwidgetui.js
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+import { ButtonView } from '@ckeditor/ckeditor5-ui/src';
+
+class ExternalWidgetUI extends Plugin {
+	init() {
+		const editor = this.editor;
+		const externalWidgetCommand = editor.commands.get( 'external' );
+
+		// The "external" button must be registered among the UI components of the editor
+		// to be displayed in the toolbar.
+		editor.ui.componentFactory.add( 'external', locale => {
+			const button = new ButtonView( locale );
+
+			button.set( {
```

Comment:
We could use the official icon. It's public domain https://en.bitcoin.it/wiki/Promotional_graphics.

Reply:
Good idea! :+1:  Added: 1712467